src/EditorFeatures/Core/Implementation/IntelliSense/SignatureHelp/Presentation/Signature.cs

```diff
@@ -16,26 +17,99 @@ internal class Signature : ISignature
         private const int MaxParamColumnCount = 100;
 
         private readonly SignatureHelpItem _signatureHelpItem;
-        private IParameter _currentParameter;
-        internal IList<SymbolDisplayPart> DisplayParts { get; private set; }
+
+        public Signature(ITrackingSpan applicableToSpan, SignatureHelpItem signatureHelpItem, int selectedParameterIndex)
+        {
+            this.ApplicableToSpan = applicableToSpan;
+            _signatureHelpItem = signatureHelpItem;
+            _parameterIndex = selectedParameterIndex;
+        }
+
+        private bool _isInitialized;
+        private void EnsureInitialized()
+        {
+            if (!_isInitialized)
+            {
+                _isInitialized = true;
+                Initialize();
+            }
+        }
+
+        private IList<SymbolDisplayPart> _displayParts;
+        internal IList<SymbolDisplayPart> DisplayParts
+        {
+            get
+            {
+                EnsureInitialized();
+                return _displayParts;
+            }
+        }
+
         private IList<SymbolDisplayPart> _prettyPrintedDisplayParts;
 
-        public ITrackingSpan ApplicableToSpan { get; internal set; }
-        public string Content { get; private set; }
-        public string PrettyPrintedContent { get; private set; }
-        public ReadOnlyCollection<IParameter> Parameters { get; private set; }
-        public string Documentation { get; private set; }
+        public ITrackingSpan ApplicableToSpan { get; }
 
-        public event EventHandler<CurrentParameterChangedEventArgs> CurrentParameterChanged;
+        private string _content;
+        public string Content
+        {
+            get
+            {
+                EnsureInitialized();
+                return _content;
+            }
+        }
 
-        public Signature(ITrackingSpan applicableToSpan, SignatureHelpItem signatureHelpItem)
+        private int _parameterIndex = -1;
+        public IParameter CurrentParameter
         {
-            this.ApplicableToSpan = applicableToSpan;
-            _signatureHelpItem = signatureHelpItem;
-            this.Initialize(setParameters: true);
+            get
+            {
+                EnsureInitialized();
+                return _parameterIndex >= 0 && _parameters != null ? _parameters[_parameterIndex] : null;
```

Comment:
If it can, I'm prepared to take the IndexOutOfRangeException. I can add validation in the constructor.